src/H5VLcallback.c

```diff
@@ -3373,8 +3476,39 @@ H5VL_file_open(const H5VL_connector_prop_t *connector_prop, const char *name, un
         HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, NULL, "not a VOL connector ID")
 
     /* Call the corresponding internal VOL routine */
-    if (NULL == (ret_value = H5VL__file_open(cls, name, flags, fapl_id, dxpl_id, req)))
-        HGOTO_ERROR(H5E_VOL, H5E_CANTOPENOBJ, NULL, "open failed")
+    if (NULL == (ret_value = H5VL__file_open(cls, name, flags, fapl_id, dxpl_id, req))) {
```

Comment:
I'm 98% on board here, but am concerned about the application specifying a particular VOL connector to open the file with, then that file open fails, and we tear off around the file system trying out VOL connectors to see if one works out.   What do you think about only iterating through the plugin path if the VOL connector class for the failed file open was the default connector?  (i.e. the native connector)

Reply:
Hmm, what if the application specifies a VOL connector to use and it's the same as the library's default connector? Should we still avoid iterating through the plugin path then? And should we differentiate that by only doing the iteration when H5P_DEFAULT is passed for the FAPL ID?

I think I wrote this up in the way you're describing for the tools. If a VOL connector ID/name is passed on the command-line, the tools should fail right away if they can't open the file with that connector. Otherwise, they try native first and then whatever is in the plugin path.

Reply:
I'm fine with either way of determining "the user didn't specify a particular VOL connector".  Whichever is easier / more sensible to you.   (Glad the tools work that way, too :-) )

Reply:
I think that the tools way is a good way to handle it. Failure with a specified VOL should be immediate, but calling H5Fopen without a VOL specified should trigger a VOL spam. I would include using the environment variable as "specifying a VOL connector", so we should only trigger the flurry of VOL checks if the default is the native VOL connector.

Reply:
The only difficult part I see is trying to differentiate if the user has actually set a VOL connector. They could pass in a FAPL that has some unrelated property set on it (for example, H5Pset_fapl_mpio might cause issues here in parallel applications), so just checking H5P_DEFAULT might be problematic here.

Reply:
This is updated now to check for H5P_DEFAULT OR compare the given connector ID against the default. This means if someone passes in a FAPL that they specifically set the default VOL connector on, it will still trigger this feature, but I think that's a very unlikely case (at least currently).